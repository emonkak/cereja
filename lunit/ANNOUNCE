
                 Lunit Release 0.3 (alpha)
                ===========================

Lunit is a unit testing framework for lua, written in lua. 

Lunit provides 26 assert functions, and a few misc functions
for usage in an easy unit testing framework.

Lunit comes with a test suite to test itself. The testsuite
consists of approximately 630 assertions.

Lunit is licensed under the MIT license.

Lunit can downloaded from http://www.nessie.de/mroth/lunit/

