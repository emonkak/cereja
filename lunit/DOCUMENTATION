
To use the unit testing framework 'lunit' copy the file 'lunit.lua' to your
lua search path or include it to your package.


To write a testcase, open the framework using require. Make sure that
require() finds the 'lunit.lua' file.

	require "lunit"


The actual tests are assertions in test functions. You can wrap a single
function to form a independet testcase. Example:

	lunit.wrap(function()
	  -- Test code goes here
	end)


Alternativly you can also supply a name for the wrapped function. Example:

	lunit.wrap("I'm a very simply test", function()
	  -- Test code goes here
	end


If you have more related tests you can group them into testcase objects. To
do this, create a new testcase object using lunit.TestCase().
Its argument is the name for the new testcase. Example:

	local testcase = lunit.TestCase("An example testcase")


The tests itself on a testcase objects are functions which names must begin
or end with 'test'. The case of these names doesn't matters. Example:

	function testcase.FirstTest()
	  -- Test code goes here
	end

	function testcase.test_something()
	    -- Test code goes here
	end


Inside the test functions, either wrapped functions or functions on a
testcase object, you use assert calls to test your code or package.
Lunit defines 26 handy assert functions:

	lunit.assert( assertion, [msg] )

		Fails, if 'assertion' is false or nil.

	lunit.assert_fail( [msg] )

		Always fails.

	lunit.assert_true( actual, [msg] )

		Fails, if 'actual' isn't true.

	lunit.assert_false( actual, [msg] )

		Fails, if 'actual' isn't false. (Even fails if 'actual' is
		a nil value!)

	lunit.assert_equal( expected, actual, [msg] )

		Fails, if 'actual' is different from 'expected'. Make sure
		that you don't mix 'expected' and 'actual' because they are
		used to build a nice error message.

	lunit.assert_not_equal( unexpected, actual, [msg] )

		Fails, if 'actual' and 'unexpected' are equal.

	lunit.assert_match( pattern, actual, [msg] )

		Fails, if the string 'actual' doesn't match 'pattern'.

	lunit.assert_not_match( pattern, actual, [msg] )

		Fails, if the string 'actual' match 'pattern'.

	lunit.assert_nil( actual, [msg] )

		Fails, if 'actual' isn't a nil value.

	lunit.assert_not_nil( actual, [msg] )

		Fails, if 'actual' is a nil value.

	lunit.assert_boolean( actual, [msg] )

		Fails, if 'actual' isn't true or false.

	lunit.assert_not_boolean( actual, [msg] )

		Fails, if 'actual' is true or false.

	lunit.assert_number( actual, [msg] )

		Fails, if 'actual' isn't a number.

	lunit.assert_not_number( actual, [msg] )

		Fails, if 'actual' is a number.

	lunit.assert_string( actual, [msg] )

		Fails, if 'actual' isn't a string.

	lunit.assert_not_string( actual, [msg] )

		Fails, if 'actual' is a string.

	lunit.assert_table( actual, [msg] )

		Fails, if 'actual' isn't a table.

	lunit.assert_not_table( actual, [msg] )

		Fails, if 'actual' is a table.

	lunit.assert_function( actual, [msg] )

		Fails, if 'actual' isn't a function.

	lunit.assert_not_function( actual, [msg] )

		Fails, if 'actual' is a function.

	lunit.assert_thread( actual, [msg] )

		Fails, if 'actual' isn't a thread (created by
		coroutine.create or coroutine.wrap).

	lunit.assert_not_thread( actual, [msg] )

		Fails, if 'actual' is a thread.

	lunit.assert_userdata( actual, [msg] )

		Fails, if 'actual' isn't userdata.

	lunit.assert_not_userdata( actual, [msg] )

		Fails, if 'actual' is userdata.

	lunit.assert_error( [msg], func )

		Fails, if 'func' doesn't raises an error (using error()).

	lunit.assert_pass( [msg], func )

		Fails, if 'func' raises an error.


All assert functions take an optional message as the last argument. Only 
assert_pass() and assert_error() require the optional message as the first
argument. The last argument of these two are functions.

There are also some handy functions to test the type of a value:

	lunit.is_nil( actual )
	lunit.is_boolean( actual )
	lunit.is_number( actual )
	lunit.is_string( actual )
	lunit.is_table( actual )
	lunit.is_function( actual )
	lunit.is_thread( actual )
	lunit.is_userdata( actual )

These all return true if 'actual' is of correct type, otherwise false. 

You use the assert functions and the is_type functions in your tests to check
your code or package. Example:

	function testcase.FirstTest()
	  local result = compute_some_value()
	  lunit.assert_string( result )
	  lunit.assert_equal("foobar", result)
	end

	function testcase.test_something()
	  local result = flip_coin()	-- flip_coin returns at random 0 or 1
	  lunit.assert_number(result)
	  if result == 0 then
	    -- ok
	  elseif result == 1 then
	    -- ok
	  else
	    lunit.assert_fail("flip_coin: invalid number: "..tostring(result))
	  end
	end


You can define the functions setup() and teardown() on testcase objects
if you have to allocate some resources or obtain some handles for your
tests. The setup() function is called before every test and teardown() is
called after every test. All tests, setup() and teardown() are called with
the test case object as the first argument. So you can add instance
variables to the object. Example:

	local tc = lunit.TestCase("Resource Test")
	
	function tc:setup()
	  self.content = { "row 1", "row 2", "row 3" }
	  self.handle = database_open("test.db")
	  database_create_table(self.handle, ...)
	  database_fill_table(self.handle, self.content, ...)
	end
	
	function tc:teardown()
	  database_drop_table(self.handle, ...)
	  database_close(self.handle)
	  self.handle = nil
	  delete_file("test.db")
	end
	
	function tc:test_select()
	  local content = database_select(self.handle, ...)
	  lunit.assert_table( content )
	  lunit.assert_equal( self.content, content )
	end
	
	function tc:test_insert()
	  database_insert(self.handle, "row 4", ...)
	  local content = database_select(self.handle, ...)
	  lunit.assert_table( content )
	  lunit.assert_equal( { "row 1", "row 2", "row 3", "row 4" }, content )
	end
	
	function tc:test_delete()
	  database_delete(self.handle, "row 2", ...)
	  local content = database_select(self.handle, ...)
	  lunit.assert_table( content )
	  lunit.assert_equal( { "row 1", "row 3" }, content )
	end


If you don't would like to type 'lunit.' before every assert or type test
function, you could import the functions using lunit.import(). Example:

	lunit.import "assert_equal"
	lunit.import "is_function"

Additionally lunit.import() understands some group names:

	lunit.import "asserts"	-- Imports all assert functions
	lunit.import "checks"	-- Imports all is_type functions
	lunit.import "all"	-- TestCase, asserts and is_type functions

You could not import lunit.run() nor lunit.import(). Only the assert
functions, the type check functions and the testcase create function
TestCase().


Calling lunit.import() installs the function to your current global
environment. If you don't would like to polute the global environment,
because this will conflict with a package under test, you could install
a private environment by calling lunit.setprivfenv(). Example:

	lunit.setprivfenv()

The call to lunit.setprivfenv() creates a new environment using setfenv()
and installs a meta table on the new environment to make sure that all
globals are accessibly.

The easiest way to use the lunit framework in your tests without global
namespace pollution is this sequence in your code:

	lunit.setprivfenv()
	lunit.import "all"

After these two calls you could use the lunit framework functions without
typing the 'lunit.' prefix and without worring about global environment
pollution.


To run all your testcases, simply call lunit.run() when you created all 
testcases and defined all tests. Example:

	lunit.run()

Testcases will be run in the order they are created calling lunit.wrap()
or lunit.TestCase(). Tests are run in the order they are created on the
testcase object.

