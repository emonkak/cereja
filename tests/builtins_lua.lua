-- Test: Built-in API (Lua part)
--
-- Copyright (C) 2006 kana <http://nicht.s8.xrea.com/>
-- $Id$

package.path = '.\\lunit\\?.lua;' .. package.path
require('lunit')

local function printf(format, ...)
  print(string.format(format, ...))
end

local function assert_error(f, ...)
  local args = {...}
  lunit.assert_error(function () f(unpack(args)) end)
end

local function assert_pass(f, ...)
  local args = {...}
  lunit.assert_pass(function () f(unpack(args)) end)
end








-- Patch part =================================================================

lunit.wrap("os.pathexistsp", function ()
  lunit.assert_true(os.pathexistsp("."));
  lunit.assert_true(os.pathexistsp(".\\libcereja.lua"));
  lunit.assert_false(os.pathexistsp("No such file or directory"));
end)




lunit.wrap("package.findfile", function ()
  local function check(expected, basename, path)
    local result = package.findfile(basename, path)
    lunit.assert_equal(expected, result)
  end

  -- ordinary cases
  check('src\\api.c', 'api', 'src\\?.c;tests\\?.c');
  check('tests\\api.c', 'api', 'tests\\?.c;src\\?.c');

  -- not found case
  check(false, 'No such file or directory', 'tests\\?.c;src\\?.c');

  -- rare cases
  check('.\\src\\api.c', 'src\\api.c', '.\\?');  -- just 1 dir in path
  check(false, 'src\\api.c', '');  -- no dir in path
end)




lunit.wrap("bitwise operations", function ()
  -- bnot) 0...0001
  -- --------------
  --       1...1110
  lunit.assert_equal(-0x2, bnot(0x1))

  --      0011
  -- bor) 0101
  -- ---------
  --      0111
  lunit.assert_equal(0x7, bor(0x3, 0x5))

  --       0011
  -- band) 0101
  -- ---------
  --       0001
  lunit.assert_equal(0x1, band(0x3, 0x5))

  --       0011
  -- bxor) 0101
  -- ---------
  --       0110
  lunit.assert_equal(0x6, bxor(0x3, 0x5))

  --          0101
  -- blshift)    1
  -- -------------
  --          1010
  lunit.assert_equal(0xA, blshift(0x5, 0x1))

  --          1010
  -- brshift)    1
  -- -------------
  --          0101
  lunit.assert_equal(0x5, brshift(0xA, 0x1))
end)




lunit.wrap("range", function ()
  local t

  t = range(0, 3)
  lunit.assert_equal(3, #t)
  lunit.assert_equal(0, t[1])
  lunit.assert_equal(1, t[2])
  lunit.assert_equal(2, t[3])

  t = range(3, 7)
  lunit.assert_equal(4, #t)
  lunit.assert_equal(3, t[1])
  lunit.assert_equal(4, t[2])
  lunit.assert_equal(5, t[3])
  lunit.assert_equal(6, t[4])

  t = range(9, 7)
  lunit.assert_equal(0, #t)

  t = range(5)
  lunit.assert_equal(5, #t)
  lunit.assert_equal(0, t[1])
  lunit.assert_equal(1, t[2])
  lunit.assert_equal(2, t[3])
  lunit.assert_equal(3, t[4])
  lunit.assert_equal(4, t[5])

  t = range(-2)
  lunit.assert_equal(0, #t)
end)


lunit.wrap("mapx", function ()
  local t, u

  t = range(1, 3+1)
  lunit.assert_equal(3, #t)
  lunit.assert_equal(1, t[1])
  lunit.assert_equal(2, t[2])
  lunit.assert_equal(3, t[3])

  u = mapx(function (v) return v * v end, t)
  lunit.assert_equal(3, #u)
  lunit.assert_equal(1, u[1])
  lunit.assert_equal(4, u[2])
  lunit.assert_equal(9, u[3])
  lunit.assert_equal(t, u)
end)


lunit.wrap("sortx", function ()
  local t, u

  t = {5, -2, 3, -4, 1}
  u = sortx(t, function (v) return (v < 0 and -v) or v end)
  lunit.assert_equal(5, #u)
  lunit.assert_equal(1, u[1])
  lunit.assert_equal(-2, u[2])
  lunit.assert_equal(3, u[3])
  lunit.assert_equal(-4, u[4])
  lunit.assert_equal(5, u[5])
  lunit.assert_equal(t, u)
end)








-- cereja part ================================================================

lunit.wrap("exported names", function ()
  lunit.assert_table(cereja)

  local keys = {}
  for k, v in pairs(cereja) do
    table.insert(keys, k)
  end
  table.sort(keys)

  for i = 1, #keys do
    printf("%s\t%s", tostring(keys[i]), type(cereja[keys[i]]))
  end
end)




lunit.wrap("cereja.load_extension", function ()
  package.cpath = ".\\?.dll" .. ";" .. package.cpath

  -- non-existent extension
  assert_error(cereja.load_extension, "foo0_bar1_baz2");

  -- fail to load
  namespace = {mode="failure", flag=false}
  assert_error(cereja.load_extension, ".\\tests\\builtins_lua_dll", namespace)

  -- success
  local function ord(c) return string.byte(c) end
  namespace = {mode="success", flag=100}
  assert_pass(cereja.load_extension, ".\\tests\\builtins_lua_dll", namespace)
  lunit.assert_equal(100 + (ord('w')+ord('a')+ord('x')), namespace.flag)

  -- load again, but nothing happend
  assert_pass(cereja.load_extension, ".\\tests\\builtins_lua_dll", namespace)
  lunit.assert_equal(100 + (ord('w')+ord('a')+ord('x')), namespace.flag)
end)








lunit.run()

-- __END__
-- vim: et sts=2 sw=2
