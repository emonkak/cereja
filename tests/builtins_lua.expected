
#### Test Suite with 8 Tests in 8 Test Cases loaded.

#### Running 'os.pathexistsp' (1 Tests)...

#### Running 'package.findfile' (1 Tests)...

#### Running 'bitwise operations' (1 Tests)...

#### Running 'range' (1 Tests)...

#### Running 'mapx' (1 Tests)...

#### Running 'sortx' (1 Tests)...

#### Running 'exported names' (1 Tests)...
NOTICE_DEBUG	number
NOTICE_ERROR	number
NOTICE_INFO	number
NOTICE_WARNING	number
_load_extension	function
_loaded_extensions	table
add_notice_handler	function
call_notice_handlers	function
exit	function
load_extension	function
notice	function
notice_debug	function
notice_debugf	function
notice_error	function
notice_errorf	function
notice_info	function
notice_infof	function
notice_warning	function
notice_warningf	function
noticef	function
on_other_shellp	boolean
remove_notice_handler	function
system_directory	string
user_directory	string

#### Running 'cereja.load_extension' (1 Tests)...

#### Test Suite finished.

54 Assertions checked. All Tests passed!
